scylla-cql/src/types/deserialize/value.rs

```diff
@@ -0,0 +1,1528 @@
+                "Expected list or set, got {:?}",
+                typ,
+            ))),
+        }
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        let v = ensure_not_null_slice(v)?;
+        let mut mem = v.as_slice();
+        let count = types::read_int_length(&mut mem)?;
+        let elem_typ = match typ {
+            ColumnType::List(elem_typ) | ColumnType::Set(elem_typ) => elem_typ,
+            _ => {
+                return Err(ParseError::BadIncomingData(format!(
+                    "Expected list or set, got {:?}",
+                    typ,
+                )))
+            }
+        };
+        Ok(Self::new(
+            elem_typ,
+            count,
+            FrameSlice::new_subslice(mem, v.as_bytes_ref()),
+        ))
+    }
+}
+
+impl<'frame, T> Iterator for SequenceIterator<'frame, T>
+where
+    T: DeserializeCql<'frame>,
+{
+    type Item = Result<T, ParseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let raw = self.raw_iter.next()?;
+        Some(raw.and_then(|raw| T::deserialize(self.elem_typ, raw)))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.raw_iter.size_hint()
+    }
+}
+
+impl<'frame, T> DeserializeCql<'frame> for Vec<T>
+where
+    T: DeserializeCql<'frame>,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        SequenceIterator::<'frame, T>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        SequenceIterator::<'frame, T>::deserialize(typ, v)?.collect()
+    }
+}
+
+impl<'frame, T> DeserializeCql<'frame> for BTreeSet<T>
+where
+    T: DeserializeCql<'frame> + Ord,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        SequenceIterator::<'frame, T>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        SequenceIterator::<'frame, T>::deserialize(typ, v)?.collect()
+    }
+}
+
+impl<'frame, T, S> DeserializeCql<'frame> for HashSet<T, S>
+where
+    T: DeserializeCql<'frame> + Eq + Hash,
+    S: BuildHasher + Default + 'frame,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        SequenceIterator::<'frame, T>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        SequenceIterator::<'frame, T>::deserialize(typ, v)?.collect()
+    }
+}
+
+/// An iterator over a CQL map.
+pub struct MapIterator<'frame, K, V> {
+    k_typ: &'frame ColumnType,
+    v_typ: &'frame ColumnType,
+    raw_iter: FixedLengthBytesSequenceIterator<'frame>,
+    phantom_data_k: std::marker::PhantomData<K>,
+    phantom_data_v: std::marker::PhantomData<V>,
+}
```

Comment:
I think `MapIterator` should have a public constructor like `SequenceIterator` or `UdtIterator`. This will be also helpful for the new API integration in cpp-rust bindings.